src/symbol.c

```diff
@@ -0,0 +1,167 @@
+}
+
+SEXP rlang_unescape_character(SEXP chr) {
+  R_xlen_t len = Rf_xlength(chr);
+  R_xlen_t i = unescape_character_in_copy(R_NilValue, chr, 0);
+  if (i == len) return chr;
+
+  SEXP ret = PROTECT(Rf_allocVector(STRSXP, len));
+  copy_character(ret, chr, i);
+  unescape_character_in_copy(ret, chr, i);
+  UNPROTECT(1);
+  return ret;
+}
+
+// Private functions -----------------------------------------------------------
+
+bool has_unicode_escape(const char* chr);
+int unescape_char(char* chr);
+int unescape_char_found(char* chr);
+int process_byte(char* tgt, char* const src, int* len_processed);
+bool has_codepoint(const char* src);
+bool is_hex(const char chr);
+
+void copy_character(SEXP tgt, SEXP src, R_xlen_t len) {
+  for (int i = 0; i < len; ++i) {
+    SET_STRING_ELT(tgt, i, STRING_ELT(src, i));
+  }
+}
+
+R_xlen_t attribute_hidden unescape_character_in_copy(SEXP tgt, SEXP src, R_xlen_t i) {
+  R_xlen_t len = Rf_length(src);
+  int dry_run = Rf_isNull(tgt);
+
+  for (; i < len; ++i) {
+    SEXP old_elt = STRING_ELT(src, i);
+    SEXP new_elt = unescape_sexp(old_elt);
+    if (dry_run) {
+      if (old_elt != new_elt) return i;
+    }
+    else {
+      SET_STRING_ELT(tgt, i, new_elt);
+    }
+  }
+
+  return i;
+}
+
+SEXP attribute_hidden unescape_sexp(SEXP name) {
+  int ce = Rf_getCharCE(name);
+  const char* src = CHAR(name);
+  const char* re_enc = Rf_reEnc(src, ce, CE_UTF8, 0);
+
+  char* tmp;
+  if (re_enc != src) {
+    // If the string has been copied, it's safe to use as buffer
+    tmp = (char*)re_enc;
+  }
+  else {
+    // If not, we're in a UTF-8 locale
+    // Need to check first if the string has any UTF-8 escapes
+    if (!has_unicode_escape(src)) return name;
+    int orig_len = strlen(re_enc);
+    tmp = alloca(orig_len + 1);
+    memcpy(tmp, re_enc, orig_len + 1);
+  }
+
+  int len = unescape_char(tmp);
+  return Rf_mkCharLenCE(tmp, len, CE_UTF8);
+}
+
+bool attribute_hidden has_unicode_escape(const char* chr) {
+  while (*chr) {
+    if (has_codepoint(chr)) {
+      return true;
+    }
+    ++chr;
+  }
+
+  return false;
+}
+
+int attribute_hidden unescape_char(char* chr) {
+  int len = 0;
+
+  while (*chr) {
+    if (has_codepoint(chr)) {
+      return len + unescape_char_found(chr);
+    }
+    else {
+      ++chr;
+      ++len;
+    }
+  }
+
+  return len;
+}
+
+int attribute_hidden unescape_char_found(char* chr) {
+  char* source = chr;
+  char* target = chr;
+  int len = 0;
+
+  while (*source) {
+    int len_processed;
+    int len_new = process_byte(target, source, &len_processed);
+    source += len_processed;
+    target += len_new;
+    len += len_new;
+  }
+
+  *target = 0;
+  return len;
+}
+
+int attribute_hidden process_byte(char* tgt, char* const src, int* len_processed) {
+  if (!has_codepoint(src)) {
+    // Copy only the first character (angle bracket or not), advance
+    *tgt = *src;
+    *len_processed = 1;
+    return 1;
+  }
+
+  int codepoint = strtol(src + strlen("<U+"), NULL, 16);
```

Comment:
I'd have thought this should be a `wchar_t` instead of an `int`, but it looks like `Rf_ucstoutf8()` takes unsigned ints. This should at least be unsigned otherwise we lose one bit.